app/controllers/admin/users_controller.rb

```diff
@@ -0,0 +1,45 @@
+class Admin::UsersController < Admin::BaseController
+  before_action :set_user, only: %i[show edit update destroy soft_delete]
+  
+  def index
+    @users = User.all
+  end
+
+  def show
+  end
+
+  def edit
+  end
+
+  def update
+    if @user.update(user_params)
+      redirect_to admin_users_path, notice: "User #{@user.email} was successfully updated."
+    else
+      render :edit, status: :unprocessable_entity 
+    end
+  end
+
+  def destroy
+    if @user.destroy
+      redirect_to admin_users_path, notice: "User #{@user.email} was successfully destroyed."
+    else
+      render :new, status: :unprocessable_entity 
+    end
+  end
+
+  def soft_delete
```

Comment:
There is no need to have a separate action for soft deletion. Instead, soft delete in destroy action and show only those users in the list whose deleted at is not set